File: apps/firebase_functions/functions/src/models/BuildJob.ts
import type { FieldValue } from "firebase-admin/firestore";

export interface BuildJob {
	buildStatus: OpenCIGitHubChecksStatus;
	github: OpenCIGithub;
	id: string;
	workflowId: string;
	createdAt: FieldValue;
}

export enum OpenCIGitHubChecksStatus {
	QUEUED = "queued",
	IN_PROGRESS = "inProgress",
	FAILURE = "failure",
	SUCCESS = "success",
}

export interface OpenCIGithub {
	repositoryUrl: string;
	owner: string;
	repositoryName: string;
	installationId: number;
	appId: number;
	checkRunId: number;
	issueNumber?: number;
	baseBranch: string;
	buildBranch: string;
}

File: packages/openci_models/lib/src/job/build_job.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'build_job.freezed.dart';
part 'build_job.g.dart';

enum OpenCIGitHubChecksStatus {
  queued,
  inProgress,
  failure,
  success,
}

@freezed
abstract class BuildJob with _$BuildJob {
  const factory BuildJob({
    required OpenCIGitHubChecksStatus buildStatus,
    required OpenCIGithub github,
    required String id,
    required String workflowId,
    @Default(null) int? createdAt,
  }) = _BuildJob;

  factory BuildJob.fromJson(Map<String, dynamic> json) =>
      _$BuildJobFromJson(json);
}

@freezed
abstract class OpenCIGithub with _$OpenCIGithub {
  OpenCIGithub._();
  const factory OpenCIGithub({
    required String owner,
    required String repositoryName,
    required int installationId,
    required int appId,
    required int checkRunId,
    @Default(null) int? issueNumber,
    required String baseBranch,
    required String buildBranch,
  }) = _OpenCIGithub;

  factory OpenCIGithub.fromJson(Map<String, dynamic> json) =>
      _$OpenCIGithubFromJson(json);

  String get repoFullName => 'https://github.com/$owner/$repositoryName';
}
